SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/WorkerCoordinator/LifetimeComponent.cs

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+using System;
+using System.Threading;
+using System.Collections.Generic;
+
+namespace Improbable.WorkerCoordinator
+{
+    /// <summary>
+    /// Simulated player client's information.
+    /// The coordinator use this to start & restart simulated player clients.
+    /// </summary>
+    struct ClientInfo
+    {
+        public string ClientName;
+        public string DevAuthToken;
+        public string TargetDeployment;
+        public long StartTick;
+        public long EndTick;
+    }
+
+    internal interface ILifetimeComponentHost
+    {
+        void StartClient(ClientInfo clientInfo);
+        void StopClient(ClientInfo clientInfo);
+        bool CheckPlayerStatus();
+    }
+
+    internal class LifetimeComponent
+    {
+        // Arguments for lifetime management.
+        public const string MaxLifetimeArg = "max_lifetime";
+        public const string MinLifetimeArg = "min_lifetime";
+        public const string RestartAfterSecondsArg = "restart_after_seconds";
+        public const string UseNewSimulatedPlayerArg = "use_new_simulated_player";
+
+        // Lifetime management parameters.
+        private bool IsLifetimeMode;
+        private ILifetimeComponentHost Host;
+        private bool UseNewSimulatedPlayer;
+        private int MaxLifetime;
+        private int MinLifetime;
+        private int RestartAfterSeconds;
+        private List<ClientInfo> WaitingList;
+        private List<ClientInfo> RunningList;
+
+        /// <summary>
+        /// Tick interval in milliseconds
+        /// </summary>
+        private int TickInterval = 1000;
+        private Logger Logger;
+        private Random Random;
+
+        // Temp variables to avoid allocate variable in tick method.
+        private long CurTicks;
```

Comment:
Why do you want to avoid allocating these? It doesn't have a cost AFAIK, but having these defined in the class would certainly increase its size.

Reply:
Oh, these are primitive types. I will move it in.